osu.Game/Screens/Play/HUD/SpectatorList.cs

```diff
@@ -104,22 +107,55 @@ protected override void LoadComplete()
             this.FadeInFromZero(200, Easing.OutQuint);
         }
 
-        private void onSpectatorsChanged(object? sender, NotifyCollectionChangedEventArgs e)
+        private void onWatchingUsersChanged(object? sender, NotifyCollectionChangedEventArgs e)
         {
+            switch (e.Action)
+            {
+                case NotifyCollectionChangedAction.Add:
+                {
+                    for (int i = 0; i < e.NewItems!.Count; i++)
+                        actualSpectators.Add((SpectatorUser)e.NewItems![i]!);
+
+                    break;
+                }
+
+                case NotifyCollectionChangedAction.Remove:
+                {
+                    for (int i = 0; i < e.OldItems!.Count; i++)
+                        actualSpectators.Remove((SpectatorUser)e.OldItems![i]!);
+
+                    break;
+                }
+
+                case NotifyCollectionChangedAction.Reset:
+                {
+                    actualSpectators.Clear();
+                    break;
+                }
+
+                default:
+                    throw new NotSupportedException();
+            }
+
+            removePlayersFromMultiplayerRoom();
+        }
+
+        private void removePlayersFromMultiplayerRoom()
+        {
+            if (multiplayerClient.Room == null)
+                return;
+
             // the multiplayer gameplay leaderboard relies on calling `SpectatorClient.WatchUser()` to get updates on users' total scores.
             // this has an unfortunate side effect of other players showing up in `SpectatorClient.WatchingUsers`.
             //
             // we do not generally wish to display other players in the room as spectators due to that implementation detail,
             // therefore this code is intended to filter out those players on the client side.
-            //
-            // note that the way that this is done is rather specific to the multiplayer use case and therefore carries a lot of assumptions
-            // (e.g. that the `MultiplayerRoomUser`s have the correct `State` at the point wherein they issue the `WatchUser()` calls).
-            // the more proper way to do this (which is by subscribing to `WatchingUsers` and `RoomUpdated`, and doing a proper diff to a third list on any change of either)
-            // is a lot more difficult to write correctly, given that we also rely on `BindableList`'s collection changed event arguments to properly animate this component.
-            var excludedUserIds = new HashSet<int>();
-            if (multiplayerClient.Room != null)
-                excludedUserIds.UnionWith(multiplayerClient.Room.Users.Where(u => u.State != MultiplayerUserState.Spectating).Select(u => u.UserID));
+            var excludedUserIds = multiplayerClient.Room.Users.Where(u => u.State != MultiplayerUserState.Spectating).Select(u => u.UserID).ToHashSet();
```

Comment:
Have you considered binding to `MultiplayerClient.CurrentMatchPlayingUserIds` instead? That is used for the spectate path so it's easier (for me) to assume it'd work here too:

https://github.com/ppy/osu/blob/f1c49dea59592a65d8ea42cac20d4f4b53b5822e/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs#L427-L443